posts/coding-intentionally-in-bash-grains.md

```diff
@@ -0,0 +1,194 @@
+-----------------------------
+```
+
+1 + 2 + 4 + 8 + 16 + 32 +64 = 127.  Do you see it?  Anything ringing any alarms with the values 31, 63, 127?
+
+They're *almost* powers of 2.  In fact, they're *one less* than the *next* power of two.  
+
+One more example, to drive it home.  Imagine a 12-square chess board. That's twelve doublings, or 2 multiplied by itself 12 times (which, in the math biz, is 2^12): 4096. Double that again and you get 8192 (2^13). So... if we got the pattern right, the running total would be *one short* of 8192, also known as 8191. And if we tally it up, that's, exactly what we get: 1 + 2 + 4 + 8 + 16 + 32 + 64 + 128 + 256 + 512 + 1024 + 2048 + 4096 = 8191.
+
+> To put it another way, to find the total for all `n` squares, all you need is the value on square `n + 1` minus one.
+
+Soooooo, if we want to total up all of the squares up through square 64, all we need to do is calculate the number of grains on *square 65* and subtract 1.
+
+Blammo.
+
+In Bash, to calculate the value on any particular square according to our new pattern, it will look like this:
+
+```bash
+bc <<< "2^($square - 1)"
+```
+
+More specifically, for the "65th" square, that would look like this:
+
+```bash
+bc <<< "2^64"
+```
+
+And to calculate the total of all of the squares that come before it (all 64 of our chessboard squares), all that's left is to subtract 1!
+
+```bash
+bc <<< "2^64 - 1"
+```
+
+This makes sense when you confirm what's happening with binary.  In binary, what was the total of all 64 squares?
+
+```txt
+0b1111...  # 64 ones
+```
+
+What's the number of grains on the theoretical 65th square?
+
+```txt
+0b10000... # 1 and 64 zeros
+```
+
+How do you get from 1 and 64 zeros down to 64 ones?  You subtract 1.
+
+And what additional benefit does this give us?  Well, now we've got a nice, readable expression for the total.  It doesn't iterate, so performance is good.  And it contains the number 64, which is the number of squares on a chessboard, which is a good example of well-signalled **design intent**.  If, for some reason, in 1000 years, the world standardizes on a 7x7 chessboard, that future engineer (probably using Bash 6.1) will check the script, see what you were going for, and change the 64 to a 49.  All good!
+
+## Stay intentional, my friends
+
+When working out an implementation, it's easy to throw things around and latch onto the first solution that works, and that's fine while you're exploring the problem, but once you fully understand the critical components — if you've got the time to spend giving things a good polish — make sure every algorithm, every variable name, and even your white space draw a picture of the problem, the critical requirements, and how all the pieces fit together.
+
+A *huge* thank you to @kytrinyx for all of her feedback and help with this post.  Much of the wisdom, clarity, and friendly explanations are thanks to her.  And @iHiD, who helped me get started and then published.
```

Comment:
Thank you ❤️ 💛 💚 💙 💜 🖤but let's cut this paragraph and end on intentional design!